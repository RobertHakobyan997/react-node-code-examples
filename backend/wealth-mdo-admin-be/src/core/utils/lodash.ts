import { assign, isArray, isObject, memoize, merge as _merge, mergeWith } from 'lodash';

export function merge<T, S>(target: T, source: S): T & S {
  return mergeWith(target, source, (targetValue, srcValue) => isArray(targetValue)
    ? assign(targetValue, srcValue)
    : isObject(srcValue)
      ? _merge(targetValue, srcValue)
      : srcValue
  );
}

/**
 * @description Memoized function for better performance it takes x4 less time
 */
export const mergeMemo = memoize(merge);
