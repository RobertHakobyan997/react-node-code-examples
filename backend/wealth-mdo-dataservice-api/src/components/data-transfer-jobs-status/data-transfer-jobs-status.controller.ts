import { Body, Controller, Delete, Get, Param, Post, Put } from '@nestjs/common';
import { ApiBody, ApiOkResponse, ApiParam, ApiTags } from '@nestjs/swagger';
import { Log } from 'ngpd-merceros-wealth-mdo-common-be/dist/logger/mdo-logger';
import DataTransferJobsStatus from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status';
import { PartialType } from '@nestjs/mapped-types';

import { DataTransferJobsStatusService } from './data-transfer-jobs-status.service';

export class PartialDataTransferJobsStatus extends PartialType<DataTransferJobsStatus>(DataTransferJobsStatus) {
}

@ApiTags('Data Transfer Job Statuses')
@Controller('data-transfer-jobs-status')
export class DataTransferJobsStatusController {
  constructor(private readonly service: DataTransferJobsStatusService) {
  }

  @ApiOkResponse({
    description: 'Returns all data transfer jobs status',
    type       : [ DataTransferJobsStatus ]
  })
  @Get()
  @Log('Get all data transfer jobs status')
  findAll() {
    return this.service.findAll();
  }

  @ApiOkResponse({
    description: 'Returns data transfer job status by id',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'id' })
  @Get(':id')
  @Log('Get data transfer job status')
  findById(@Param('id') id: string) {
    return this.service.findById(id);
  }

  @ApiOkResponse({
    description: 'Get files by files metadata id',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'jobMetaDataId', description: 'Metadata id' })
  @Get('metadata/:jobMetaDataId')
  @Log('Get files by job metadata id')
  getFilesByMetadataId(@Param('jobMetaDataId') jobMetaDataId: string) {
    return this.service.getJobMetadataStatuses(jobMetaDataId);
  }

  @ApiOkResponse({ description: 'Update data transfer job status by id' })
  @ApiBody({
    description: 'Data transfer job status',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'id' })
  @ApiBody({ description: 'Data transfer job status', type: DataTransferJobsStatus })
  @Put(':id')
  @Log('Update data transfer job status of existing item')
  update(
    @Param('id') id: string,
    @Body() dataTransferJobsStatus: PartialDataTransferJobsStatus,
  ) {
    return this.service.update(id, dataTransferJobsStatus);
  }

  @ApiOkResponse({ description: 'Update/Create data transfer job status by job metadata id' })
  @ApiBody({
    description: 'Data transfer job status',
    type       : DataTransferJobsStatus,
  })
  @ApiParam({ name: 'jobMetaDataId', description: 'Metadata id' })
  @ApiBody({ description: 'Data transfer job status', type: DataTransferJobsStatus })
  @Put('metadata/:jobMetaDataId')
  @Log('Update data transfer job status of existing item or create the new one')
  updateJobMetadataStatuses(
    @Param('jobMetaDataId') id: string,
    @Body() dataTransferJobsStatus: Partial<DataTransferJobsStatus>,
  ) {
    return this.service.updateJobMetadataStatuses(id, dataTransferJobsStatus);
  }

  @ApiOkResponse({
    description: 'Creates data transfer job status',
    type       : [ DataTransferJobsStatus ],
  })
  @ApiBody({ description: 'New data transfer job status', type: [ DataTransferJobsStatus ] })
  @Post()
  @Log('Create data transfer job status')
  create(
    @Body() dataTransferJobsStatus: DataTransferJobsStatus[]
  ) {
    return this.service.create(dataTransferJobsStatus);
  }

  @ApiOkResponse({
    description: 'Deletes data transfer job status by id',
    type       : [ DataTransferJobsStatus ],
  })
  @ApiParam({ name: 'id' })
  @Delete(':id')
  @Log('Delete data transfer job status')
  delete(@Param('id') id: string) {
    return this.service.delete(id);
  }
}
