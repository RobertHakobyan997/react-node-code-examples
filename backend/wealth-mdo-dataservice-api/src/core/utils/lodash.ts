import {
  assign,
  isArray,
  isObject,
  memoize,
  merge as _merge,
  mergeWith,
} from 'lodash';

export function merge<T, S>(target: T, source: S): T & S {
  return mergeWith(target, source, (targetValue, srcValue) => {
    if (isArray(targetValue)) {
      // assign only array of entities, return src if array of primitives
      if (srcValue.every(s => isObject(s))) {
        return assign(targetValue, srcValue);
      }
      return srcValue;
    }
    if (isObject(srcValue)) {
      return _merge(targetValue, srcValue);
    }
    return srcValue;
  });
}

/**
 * @description Memoized function for better performance it takes x4 less time
 */
export const mergeMemo = memoize(merge);
