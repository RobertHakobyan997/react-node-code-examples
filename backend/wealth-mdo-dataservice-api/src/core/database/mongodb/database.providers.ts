import mongoose from 'mongoose';
import { Logger } from '@nestjs/common';
import { IS_LOCAL } from 'ngpd-merceros-wealth-mdo-common-be/dist/utils/environments';

import mongoConfig from '../../../config/mongo.config';

export const databaseProviders = [
  {
    // ToDo: change to dynamic provider and take it from mongoConfig
    provide   : 'DatabaseConnection',
    useFactory: (): Promise<typeof mongoose> => {
      if (IS_LOCAL()) {
        mongoose.set('debug', (collectionName, methodName, ...methodArgs) => {
          const logger = new Logger(collectionName);
          const args = methodArgs.reduce((acc, next, idx) => {
            acc += JSON.stringify(next, null, 2) + (idx === methodArgs.length - 1 ? '' : ', ');
            return acc;
          }, '');
          logger.log(`${methodName}(${args})`);
        });
      }
      return mongoose.connect(mongoConfig().uri, mongoConfig().options);
    },
  },
];
