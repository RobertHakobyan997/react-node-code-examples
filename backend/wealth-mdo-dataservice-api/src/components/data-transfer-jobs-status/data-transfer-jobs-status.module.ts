import { Module } from '@nestjs/common';
import DataTransferJobsStatusModel, { DataTransferJobsStatusModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-status';

import { DatabaseModule } from '../../core/database/mongodb/database.module';

import { DataTransferJobsStatusController } from './data-transfer-jobs-status.controller';
import { DataTransferJobsStatusService } from './data-transfer-jobs-status.service';
import { DataTransferJobsStatusDal } from './data-transfer-jobs-status.dal';

@Module({
  imports    : [ DatabaseModule ],
  controllers: [ DataTransferJobsStatusController ],
  providers  : [
    { provide: DataTransferJobsStatusModelToken, useFactory: () => DataTransferJobsStatusModel },
    DataTransferJobsStatusService,
    DataTransferJobsStatusDal
  ],
  exports: [ DataTransferJobsStatusService ],
})
export class DataTransferJobsStatusModule {
}
