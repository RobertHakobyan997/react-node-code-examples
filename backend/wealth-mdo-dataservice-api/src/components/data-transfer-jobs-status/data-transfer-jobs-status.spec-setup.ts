import { Test, TestingModule } from '@nestjs/testing';
import { DataTransferJobsStatusModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-status';

import { mockMongooseModel } from '../../../test/mocks/mock-mongoose.model';

import { DataTransferJobsStatusService } from './data-transfer-jobs-status.service';
import { DataTransferJobsStatusDal } from './data-transfer-jobs-status.dal';

export const assert = async () => {
  const module: TestingModule = await Test.createTestingModule({
    providers: [
      DataTransferJobsStatusService,
      DataTransferJobsStatusDal,
      { provide: DataTransferJobsStatusModelToken, useValue: mockMongooseModel },
    ],
  }).compile();

  return module.get<DataTransferJobsStatusService>(DataTransferJobsStatusService);
};
