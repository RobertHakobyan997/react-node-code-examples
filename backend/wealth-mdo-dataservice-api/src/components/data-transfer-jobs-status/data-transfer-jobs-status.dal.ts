import { Inject, Injectable } from '@nestjs/common';
import { FilterQuery, Model, UpdateQuery } from 'mongoose';
import { Logger } from 'ngpd-merceros-logger-be';
import { plainToClass } from 'class-transformer';
import DataTransferJobsStatusModel, { DataTransferJobsStatusModelToken } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/model/mongoose/data-transfer-jobs-status';
import DataTransferJobsStatus from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status';
import DataTransferJobsStatusState from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status-state';
import { Document } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/mongodb/document';
import { ObjectId } from 'mongodb';

import { mergeMemo } from '../../core/utils/lodash';

@Injectable()
export class DataTransferJobsStatusDal {
  private readonly logger = new Logger(DataTransferJobsStatusDal.name);

  constructor(@Inject(DataTransferJobsStatusModelToken) private readonly model: Model<Document<DataTransferJobsStatus>>) {
  }

  findAll(filter: FilterQuery<Document<DataTransferJobsStatus>> = {}, sort: string | any = {}, options: any = {}) {
    return this.model
      .find(filter, {}, options)
      .sort(sort)
      .exec();
  }

  findOne(filter: FilterQuery<Document<DataTransferJobsStatus>> = {}, projection: any = {}, options: any = {}) {
    return this.model.findOne(filter, projection, options).exec();
  }

  findById(id: any, options: any = {}) {
    return this.model.findById(id, {}, options).exec();
  }

  insert(docs: any[]) {
    return this.model.insertMany(plainToClass(DataTransferJobsStatusModel, docs));
  }

  async updateByJobMetaDataId(id: any, docs: UpdateQuery<DataTransferJobsStatus>, options: any = {}) {
    const prev = await this.findOne({ jobMetaDataId: new ObjectId(id) });
    const updated = mergeMemo(prev, docs);
    await this.model.findByIdAndUpdate(id, plainToClass(DataTransferJobsStatusModel, updated), options);
    return updated;
  }

  async updateOne(_id: any, docs: UpdateQuery<DataTransferJobsStatus>, options: any = {}) {
    return this.model.updateOne({ _id }, docs, options);
  }

  findByIdAndUpdate(id: any, docs: UpdateQuery<DataTransferJobsStatus>, options: any = {}) {
    return this.model.findByIdAndUpdate(id, plainToClass(DataTransferJobsStatusModel, docs), options);
  }

  findByIdAndDelete(id: any) {
    return this.model.findByIdAndDelete(id);
  }

  update(status: UpdateQuery<DataTransferJobsStatusState>, filter: FilterQuery<Document<DataTransferJobsStatus>> = {}) {
    return this.model.updateOne(filter, plainToClass(DataTransferJobsStatusModel, status));
  }
}
