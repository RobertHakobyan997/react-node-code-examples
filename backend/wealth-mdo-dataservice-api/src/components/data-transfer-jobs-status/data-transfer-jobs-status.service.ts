import { BadRequestException, Injectable } from '@nestjs/common';
import { FilterQuery } from 'mongoose';
import { ObjectId } from 'mongodb';
import DataTransferJobsStatus from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status';
import { Document } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/mongodb/document';

import { mergeMemo } from '../../core/utils/lodash';
import { defaultDate } from '../../core/utils/luxon';

import { DataTransferJobsStatusDal } from './data-transfer-jobs-status.dal';

@Injectable()
export class DataTransferJobsStatusService {
  constructor(private readonly dal: DataTransferJobsStatusDal) {
  }

  findAll(filter: FilterQuery<Document<DataTransferJobsStatus>> = {}, sort: string | any = {}, options: any = {}) {
    return this.dal.findAll(filter, sort, options);
  }

  findById(id: string) {
    return this.dal.findById(id);
  }

  getJobMetadataStatuses(jobMetaDataId: string) {
    return jobMetaDataId && ObjectId.isValid(jobMetaDataId)
      ? this.dal.findOne({ jobMetaDataId: ObjectId.createFromHexString(jobMetaDataId) }, {}, { lean: true })
      : undefined;
  }

  async update(id: string, status: Partial<DataTransferJobsStatus>) {
    const dataTransferJobsStatus = await this.findById(id);
    if (!dataTransferJobsStatus) throw new BadRequestException();

    return this.updateExistingStatus(status, dataTransferJobsStatus);
  }

  // TODO: refactor this - jpb metadata id mutates if not set in status before
  async updateJobMetadataStatuses(id: string, status: Partial<DataTransferJobsStatus>) {
    const statusWithId = {
      jobMetaDataId: id,
      state        : {
        processingStatus         : status.state.processingStatus ?? '',
        errorMessage             : status.state.errorMessage ?? '',
        lastSuccessfulProcessTime: status.state.lastSuccessfulProcessTime ?? defaultDate(),
        lastRunTime              : status.state.lastRunTime ?? defaultDate(),
        totalSuccessCount        : status.state.totalSuccessCount ?? 0,
      },
    } as unknown as DataTransferJobsStatus;
    const dataTransferJobsStatus = await this.getJobMetadataStatuses(id);
    if (!dataTransferJobsStatus) {
      return this.create([ statusWithId ]);
    }

    return this.updateExistingStatus(statusWithId, dataTransferJobsStatus);
  }

  async updateExistingStatus(target: Partial<DataTransferJobsStatus>, source: DataTransferJobsStatus) {
    const updatedStatus = mergeMemo(JSON.parse(JSON.stringify(source)), JSON.parse(JSON.stringify(target)));
    await this.dal.update(updatedStatus, { _id: source._id });
    return updatedStatus;
  }

  create(body: DataTransferJobsStatus[]) {
    return this.dal.insert(body);
  }

  delete(id: string) {
    return this.dal.findByIdAndDelete(id);
  }
}
