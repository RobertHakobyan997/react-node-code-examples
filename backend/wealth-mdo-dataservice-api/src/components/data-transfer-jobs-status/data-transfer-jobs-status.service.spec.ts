import { cloneDeep } from 'lodash';
import DataTransferJobsStatus from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/data-transfer-jobs-status';
import { Document } from 'ngpd-merceros-wealth-mdo-common-be/dist/domain/structures/mongodb/document';
import { plainToClass } from 'class-transformer';
import { ObjectId } from 'mongodb';

import { dataTransferJobsStatus, dataTransferJobsStatusRaw } from '../../../test/data/data-transfer-jobs/data-transfer-jobs-status.const';

import { DataTransferJobsStatusService } from './data-transfer-jobs-status.service';
import { assert } from './data-transfer-jobs-status.spec-setup';

describe('ProcessingStatusesService', () => {
  let service: DataTransferJobsStatusService;
  let jobStatus;

  beforeEach(async () => {
    service = await assert();
    jobStatus = cloneDeep(dataTransferJobsStatus[0]);
  });

  xit('plainToClass', () => {
    const mapped = plainToClass(DataTransferJobsStatus, dataTransferJobsStatusRaw);
    // ToDo: fix structure and entity in common
    // expect(mapped._id).toBeInstanceOf(Types.ObjectId);
    expect(mapped).toMatchSnapshot();
  });

  describe('findAll()', () => {
    it('should return all data transfer jobs status', async () => {
      jest.spyOn(service['dal'], 'findAll').mockResolvedValue(dataTransferJobsStatus as Document<DataTransferJobsStatus>[]);
      const res = await service.findAll();
      expect(res).toEqual(dataTransferJobsStatus);
      expect(res.length).toBe(1);
    });
  });

  describe('findById()', () => {
    it('should return data transfer job status by id', async () => {
      jest.spyOn(service['dal'], 'findById').mockResolvedValue(jobStatus);
      const res = await service.findById(dataTransferJobsStatus[0]._id.toString());
      expect(res).toEqual(jobStatus);
    });
  });

  describe('create()', () => {
    it('should create and return data transfer job status', async () => {
      jest.spyOn(service['dal'], 'insert').mockImplementation((data: DataTransferJobsStatus[]) => data as any);
      const res = await service.create(dataTransferJobsStatus);
      expect(res).toEqual(dataTransferJobsStatus);
      expect(service['dal'].insert).toHaveBeenCalledWith(dataTransferJobsStatus);
    });
  });

  describe('delete()', () => {
    it('should delete data transfer job status and return result of the operation', async () => {
      jest.spyOn(service['dal'], 'findByIdAndDelete').mockResolvedValue(jobStatus);
      await service.delete(jobStatus._id.toString());
      expect(service['dal'].findByIdAndDelete).toHaveBeenCalledWith(jobStatus._id.toString());
    });
  });

  describe('getJobMetadataStatuses', () => {
    it('should return job metadata statuses', async () => {
      jest.spyOn(service['dal'], 'findOne').mockResolvedValue(jobStatus);
      await service.getJobMetadataStatuses(jobStatus.jobMetaDataId.toString());
      expect(service['dal'].findOne).toHaveBeenCalledWith(
        { jobMetaDataId: new ObjectId(jobStatus.jobMetaDataId) },
        {},
        { lean: true }
      );
    });
    it('should return undefined', async () => {
      const result = await service.getJobMetadataStatuses('');
      expect(result).toBeUndefined();
    });
  });

  describe('update', () => {
    it('should update', async () => {
      const test = { state: {} } as DataTransferJobsStatus;
      jest.spyOn(service['dal'], 'findById').mockResolvedValue(jobStatus);
      jest.spyOn(service, 'updateExistingStatus').mockResolvedValue(jobStatus);
      await service.update(jobStatus._id.toString(), test);
      expect(service.updateExistingStatus).toHaveBeenCalledWith(test, jobStatus);
    });
    it('should throw exception', async () => {
      const test = { state: {} } as DataTransferJobsStatus;
      jest.spyOn(service['dal'], 'findById').mockResolvedValue(undefined);
      await expect(async() => {
        await service.update(jobStatus._id.toString(), test);
      }).rejects.toThrowError();
    });
  });

  describe('updateJobMetadataStatuses', () => {
    it('should call updateExistingStatus', async () => {
      const statusWithId = {
        jobMetaDataId: jobStatus._id.toString(),
        state        : {
          processingStatus         : jobStatus.state.processingStatus,
          errorMessage             : jobStatus.state.errorMessage,
          lastSuccessfulProcessTime: jobStatus.state.lastSuccessfulProcessTime,
          lastRunTime              : jobStatus.state.lastRunTime,
          totalSuccessCount        : jobStatus.state.totalSuccessCount,
        }
      };
      jest.spyOn(service, 'getJobMetadataStatuses').mockResolvedValue(jobStatus);
      jest.spyOn(service, 'updateExistingStatus').mockResolvedValue(jobStatus);
      await service.updateJobMetadataStatuses(jobStatus._id.toString(), jobStatus);
      expect(service.updateExistingStatus).toHaveBeenCalledWith(statusWithId, jobStatus);
    });
    it('should call create', async () => {
      const statusWithId = {
        jobMetaDataId: jobStatus._id.toString(),
        state        : {
          processingStatus         : jobStatus.state.processingStatus,
          errorMessage             : jobStatus.state.errorMessage,
          lastSuccessfulProcessTime: jobStatus.state.lastSuccessfulProcessTime,
          lastRunTime              : jobStatus.state.lastRunTime,
          totalSuccessCount        : jobStatus.state.totalSuccessCount,
        }
      };
      jest.spyOn(service, 'getJobMetadataStatuses').mockResolvedValue(undefined);
      jest.spyOn(service, 'create').mockResolvedValue(jobStatus);
      await service.updateJobMetadataStatuses(jobStatus._id.toString(), jobStatus);
      expect(service.create).toHaveBeenCalledWith([ statusWithId ]);
    });
  });

  describe('updateExistingStatus', () => {
    it('should update existing status', async () => {
      const test = { state: {} } as DataTransferJobsStatus;
      jest.spyOn(service['dal'], 'update').mockResolvedValue(jobStatus);
      await service.updateExistingStatus(test, jobStatus);
      expect(service['dal'].update).toHaveBeenCalled();
    });
  });
});
