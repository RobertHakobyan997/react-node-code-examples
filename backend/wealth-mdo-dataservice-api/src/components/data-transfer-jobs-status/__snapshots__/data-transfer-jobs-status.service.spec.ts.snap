// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessingStatusesService plainToClass 1`] = `
DataTransferJobsStatus {
  "_id": "611f68a99f217cce3a4d91f8",
  "jobMetaDataId": "611d04f5204471d32fd1c18e",
  "state": Object {
    "errorMessage": "",
    "lastRunTime": "2021-08-05T17:20:15.413Z",
    "lastSuccessfulProcessTime": "2021-08-05T17:20:30.437Z",
    "processingStatus": "technicalError",
  },
  "updatedAt": "2021-08-05T17:20:30.437Z",
}
`;
